toss/fitness/fitness_function_utils.py

```diff
@@ -45,4 +45,68 @@ def _compute_squared_distance(positions,constant):
     Returns:
         (np.ndarray): (N,) array containing average distance from each point in arr to constant.
     """
-    return np.sum(np.power(positions,2), axis=0) - constant**2
\ No newline at end of file
+    return np.sum(np.power(positions,2), axis=0) - constant**2
+
+
+def compute_space_coverage(positions, velocities, timesteps, radius_outer_bounding_sphere):
+    """
+    In this function, we create a spherical meshgrid by defining a number of points
+    inside the outer bounding sphere. We then generate a multidimensional array
+    of boolean values corresponding to each point in the meshgrid. These values 
+    are initialy set to false since they have not been visited by any spacecraft.
+
+    Given a set of posiions on the trajectory that are defined inside the 
+    outer bounding sphere, we identify the points that are the closest the given positions, 
+    and subsequently reassign the visited boolean value to true. The ratio of visited points 
+    is then given by the number of True values to the total number of values in the boolean array.
+
+    Args:
+        positions (np.ndarray): (3,N) Array of positions (in cartesian coordinates).
+        velocities (np.ndarray): (3,N) Array of positions (in cartesian frame).
+        timesteps (np.ndarray): (N) Array of time values for each position.
+        radius_outer_bounding_sphere (float): Radius of outer bounding sphere.
+
+    Returns:
+        ratio (float): Number of True values to the total number of values in the boolean array
+    """
+
+    # Define frequency of points for the spherical meshgrid: (see: Courant–Friedrichs–Lewy condition)
+    cfl = 1 # Courant number
+    v_max = np.sqrt(np.max(velocities[0,:]**2 + velocities[1,:]**2 + velocities[2,:]**2))
+    dt = timesteps[1]-timesteps[0]
+    
+    # Calculate and adjust grid spacing based on the Courant number, maximal velocity, and time step
+    grid_spacing = v_max * dt / cfl
+    r = np.linspace(0, radius_outer_bounding_sphere, int(radius_outer_bounding_sphere / grid_spacing)) # Number of evenly spaced points along the radial axis
+    theta = np.linspace(0, np.pi, int(np.pi / grid_spacing)) # Number of evenly spaced points along the polar angle (defined in between 0 and pi)
+    phi = np.linspace(0, 2 * np.pi, int(2 * np.pi / grid_spacing)) # Number of evenly spaced points along the azimuthal angle (defined in between 0 and 2*pi)
```

Comment:
I don't think this works, grid_spacing is in distance, the angles are in radiant. What I would suggest to do is compute angular distance traveled per cell by computing delta_phi=2*pi*r_min / grid_steps and then making sure v_max * dt < delta_phi. It's a bit of a hack but should be fine, I reckon.

Reply:
(also notice that you are currently dividing by the number of grid points by `v_max * dt` so the number of grid points converges to infinity for v_max = 0, wrong way round)